src/storage/txn/commands/acquire_pessimistic_lock.rs

```diff
@@ -86,12 +86,16 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for AcquirePessimisticLock
             ))));
         }
 
-        let (start_ts, ctx, keys) = (self.start_ts, self.ctx, self.keys);
+        let (start_ts, ctx, mut keys) = (self.start_ts, self.ctx, self.keys);
         let mut txn = MvccTxn::new(start_ts, context.concurrency_manager);
-        let mut reader = ReaderWithStats::new(
-            SnapshotReader::new_with_ctx(start_ts, snapshot, &ctx),
-            context.statistics,
-        );
+
+        let mut snapshot_reader = SnapshotReader::new_with_ctx(start_ts, snapshot, &ctx);
+        if keys.len() > 1 {
+            keys.sort();
+            snapshot_reader.set_scan_mode(ScanMode::Forward);
+            snapshot_reader.set_lower_bound(keys.first().unwrap().0.clone());
+        }
```

Comment:
Can we wrap the logic into `SnapshotReader::new_with_ctx` like
```
SnapshotReader::new_with_ctx(start_ts, snapshot, &ctx, &keys);
```
, so the duplicated code could be avoided.

Reply:
`keys` is not available for all cases, eg. it's `mutations` in prewrite and `locks` in resolve_lock. so we may need sth like "sort_by_key_and_get_first_key".

Reply:
Yes, there are `mutations` and `keys` actually. We can do something like this.

```rust
trait IntoKey {
    fn key(&self) -> &Key;
}

impl IntoKey for MutationLock {
    fn key(&self) -> &Key {
        self.key()
    }
}

impl IntoKey for PessimisticMutation {
    fn key(&self) -> &Key {
        self.mutation.key()
    }
}

impl IntoKey for Key {
    fn key(&self) -> &Key {
        self
    }
}
```

Reply:
How about adding the following method to `SnapshotReader`? 
```rs
    pub fn setup_with_hint_items<T>(&mut self, items: &mut Vec<T>, key_of: fn(&T) -> &Key) {
        if items.len() > 1 {
            items.sort_by(|a, b| key_of(a).cmp(&key_of(b)));
            self.set_scan_mode(ScanMode::Forward);
            self.set_lower_bound(key_of(items.first().unwrap()).clone());
        }
    }
```
Then we can call `snapshot_reader.setup_with_hint_items(&mut key_locks, |(key, _)| key);` in resolve lock.

Reply:
The `pub fn setup_with_hint_items<T>(&mut self, items: &mut Vec<T>, key_of: fn(&T) -> &Key)` looks simpler, personally.